File: passpersist-perl/ethtool-stats.pl
#!/usr/bin/perl

use strict;
use SNMP::Extension::PassPersist;

my $base = ".1.3.6.1.4.1.39178.100.1.1.1.2";

my $extsnmp = SNMP::Extension::PassPersist->new(
    backend_collect => \&update_tree
    );
$extsnmp->run;


sub update_tree {
    my ($self) = @_;

    # We grab interfaces from /sys/class/net
    my @interfaces = </sys/class/net/*>;
    foreach my $interface (@interfaces) {
        # Get index of this interface
        open(IFINDEX, "$interface/ifindex") or next;
        my $index = int(<IFINDEX>);
        close(IFINDEX);

        # Call ethtool
        $interface =~ s/^.*\///;
        open(ETHTOOL, "ethtool -S $interface 2>/dev/null |") or next;
        while (<ETHTOOL>) {
            # Extract name and value
            /^\s+(\w+): (\d+)$/ or next;
            my $name = $1;
            my $value = int($2);
            # Compute OID
            my $oid = "$base.$index";
            foreach my $char (split //, $name) {
                $oid .= ".";
                $oid .= ord($char);
            }
            # Append to our tree
            $extsnmp->add_oid_entry($oid, "counter", $value);
        }
        close(ETHTOOL);
    }
}

File: passpersist-python/ethtool-stats.py
import re
import os
import glob
import subprocess

import snmp_passpersist as snmp

def update():
    for interface in os.listdir("/sys/class/net"):
        # Get index
        index = open("/sys/class/net/%s/ifindex" % interface)
        index = int(index.read())

        # Call ethtool
        try:
            ethtool = subprocess.check_output(["ethtool", "-S", interface],
                                              stderr=subprocess.STDOUT)
        except subprocess.CalledProcessError:
            continue

        ethtool = ethtool.decode("ascii")
        for line in ethtool.split("\n"):
            mo = re.match("\s+(\w+): (\d+)", line)
            if not mo: continue
            name = mo.group(1)
            value = int(mo.group(2))
            oid = "%d.%s" % (index,
                             ".".join([str(ord(a)) for a in name]))
            pp.add_cnt_64bit(oid, value)

pp = snmp.PassPersist('.1.3.6.1.4.1.39178.100.1.1.1.2')
pp.start(update, 10)
